src/tests/test_first.cpp

```diff
@@ -0,0 +1,122 @@
+//
+//           Copyright Aleksey Loginov 2022 - present.
+//                     TC Wang 2022 - present.
+//  Distributed under the Boost Software License, Version 1.0.
+//     (See accompanying file LICENSE_1_0.txt or copy at
+//           https://www.boost.org/LICENSE_1_0.txt)
+//
+//  Project home: https://github.com/victimsnino/ReactivePlusPlus
+
+#include <catch2/catch_test_macros.hpp>
+
+#include <vector>
+
+#include <rpp/operators/first.hpp>
+#include <rpp/sources/empty.hpp>
+#include <rpp/sources/error.hpp>
+#include <rpp/sources/just.hpp>
+#include <rpp/sources/never.hpp>
+
+#include "mock_observer.hpp"
+
+SCENARIO("first only emits once", "[first]")
+{
+    GIVEN("observable of -1-|")
+    {
+        auto mock = mock_observer<int>{};
+        auto obs = rpp::source::just(1);
+        obs.first()
+            .subscribe(mock);
+        THEN("shall see -1-|")
+        {
+            CHECK(mock.get_received_values() == std::vector{1});
+            CHECK(mock.get_on_completed_count() == 1);
+            CHECK(mock.get_on_error_count() == 0);
+        }
+    }
+
+    GIVEN("observable of -1-2-3-|")
+    {
+        auto mock = mock_observer<int>{};
+        auto obs = rpp::source::just(1, 2, 3);
+        obs.first()
+            .subscribe(mock);
+        THEN("shall see -1-|")
+        {
+            CHECK(mock.get_received_values() == std::vector{1});
+            CHECK(mock.get_on_completed_count() == 1);
+            CHECK(mock.get_on_error_count() == 0);
+        }
+    }
+
+    GIVEN("observable of never")
+    {
+        auto mock = mock_observer<int>{};
+        auto obs = rpp::source::never<int>();
+        obs.first()
+            .subscribe(mock);
+        THEN("shall see ---|")
+        {
+            CHECK(mock.get_received_values().empty());
+            CHECK(mock.get_on_completed_count() == 0);
+            CHECK(mock.get_on_error_count() == 0);
+        }
+    }
+}
+
+SCENARIO("first forwards error", "[first]")
+{
+    GIVEN("observable of x-|")
+    {
+        auto mock = mock_observer<int>{};
+        auto obs = rpp::source::error<int>(std::make_exception_ptr(std::runtime_error{""}));
+        obs.first()
+            .subscribe(mock);
+        THEN("shall see error and no-completed")
+        {
+            CHECK(mock.get_received_values().empty());
+            CHECK(mock.get_on_completed_count() == 0);
+            CHECK(mock.get_on_error_count() == 1);
+        }
+    }
+}
+
+SCENARIO("first raises error for empty", "[first]")
+{
+    GIVEN("observable of ---|")
+    {
+        auto mock = mock_observer<int>{};
+        auto obs = rpp::source::empty<int>();
+        obs.first()
+            .subscribe(mock);
+        THEN("shall see -x")
+        {
+            CHECK(mock.get_received_values().empty());
+            CHECK(mock.get_on_completed_count() == 0);
+            CHECK(mock.get_on_error_count() == 1);
+        }
+    }
+}
+
+SCENARIO("first regards unsubscription", "[first]")
```

Comment:
Not sure if it checks something related to first. Subscriber would not obtain any emissions if subscription is unsubscribed

Reply:
Oh that's already taken care of in the subscriber and thus we don't need to reinforce that in lift's operator?

Reply:
Yes, absolutely. There is no way send something to "unsubscribed" subscriber. Because
1) unsubscribed subscriber can't subscribe to new observable
2) unsubscribed subscriber can't get any on_next/on_error/on_completed due to extra checks


Only one way when we need to do it: inside "infinite-like" observables. For example, if we iterate over vector, then we don't need to continue iteration in case of subscriber unsubscribed

Reply:
Moreover, if downstream and upstream subscribers use same subscription/child, then you can't even go to "operator's on_next" due to subscriber unsubcribed and it means that your "operator's subscriber" unsubscribed too and it can't accept new value

Reply:
> Moreover, if downstream and upstream subscribers use same subscription/child, then you can't even go to "operator's on_next" due to subscriber unsubcribed and it means that your "operator's subscriber" unsubscribed too and it can't accept new value

This is also a good topic to add it in my article collection.

Reply:
> Only one way when we need to do it: inside "infinite-like" observables.

Yup I got it. It makes more sense to add `is_subscribed()` check in the source.